### PR TITLE
[KEYCLOAK-9841] use LDAPUser UUID as an identifier instead of username

### PR DESCRIPTION
**Pull request by PacoVK**
Implementation and Test added according the description in KEYCLOAK-9841 to load LDAPUser by UUID and prevent orphan users in localstorage

### CODE REVIEW

**Review by mposolda — CHANGES_REQUESTED** [review]
@PacoVK Thanks for the new  PR!

The part related to "Sync users" looks good to me and also great work with the automated test. Just one small thing - is it possible to use usernames without dot like "beckybecks" instead of "becky.becks" ? I am not 100% sure if dots work with MS Active Directory and other LDAP servers, which we need to test with...

The more major thing is, that we need the complete fix for this issue before we merge this. The PR address just the scenario of syncing users by the "Sync users", but we need to cover also the "ad-hoc" syncing of users. In other words the scenario, when the user, which was renamed in LDAP, is trying to login (or is directly searched by admin by username in the admin console).

It will be ideal to make sure that LDAPStorageProvider.loadAndValidateUser doesn't try to find users by username, but by LDAP UUID. In other words, the authoritative source of information is the UUID from LDAP. Other place, which will probably require some change is the UserStorageManager.getUserByUsername . Will be ideal to treat the username similarly like for example the email and hence something like this might be needed for username too: https://github.com/keycloak/keycloak/blob/12.0.4/services/src/main/java/org/keycloak/storage/UserStorageManager.java#L264-L267

It will be ideal to have test like this working (for example in LDAPProvidersIntegrationTest class):
```
     @Test
    public void updateLDAPUsernameTest() {
        testingClient.server().run(session -> {
            LDAPTestContext ctx = LDAPTestContext.init(session);
            UserModel user = session.users().getUserByUsername(ctx.getRealm(), "beckybecks");
            Assert.assertNotNull(user);
            String origKeycloakUserId = user.getId();
        });

        // TODO: Rename user "beckybecks" to "beckyupdated" here...

        testingClient.server().run(session -> {
            LDAPTestContext ctx = LDAPTestContext.init(session);

            // The original username is not possible to use as username was changed in LDAP.
            // However the call to LDAPStorageProvider.loadAndValidateUser shouldn't delete the user just because his username changed in LDAP
            UserModel user = session.users().getUserByUsername(ctx.getRealm(), "beckybecks");
            Assert.assertNull(user);

            // Assert user can be found with new username from LDAP. And it is same user as before
            user = session.users().getUserByUsername(ctx.getRealm(), "beckyupdated");
            Assert.assertNotNull(user);            
            String newKeycloakUserId = user.getId();
            Assert.assertEquals(keycloakUserId, newKeycloakUserId);
        });
    }
```
I am sorry for the additional work, but I have a feeling that will be nice to address this properly on all places instead of single place. What do you think?

**Comment by PacoVK** [comment]
Hey @mposolda nice catch and i think there are good points you mentioned. Apparently there is more work to do and changes will be bigger. I'd like to take care about that and ensure users are loaded by UUID and update this PR once the work ist done 👍

**Comment by mposolda** [comment]
@PacoVK Cool, Thanks!

**Comment by PacoVK** [comment]
@mposolda used a bit different approach, treating UUID the same as Email is IMHO not a good solution as both email and username are valid login names for all providers. UUID is LDAP specific hence i changed the implementation for the LDAP provider to guarantee to have no duplicated/ stale users. this makes query for username safe again. However still have a test to fix, not quite sure if it is related to the testdata in UserModelTest.testAddDirtyRemoveFederationUsersInTheSameGroupConcurrent 
WDYT?

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java
Comment (mposolda) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:526]: Won't it be easier to use just `imported = existingLocalUser;` on this line? I think there is no need to trigger again `session.userLocalStorage().getUserByUsername` as that should always return same user as `existingLocalUser` if I am not mistaken? So this would just mean another SQL query to the DB.

**Review by mposolda — CHANGES_REQUESTED** [review]
@PacoVK That works for me. Thanks for adding this and also for adding the test. I have one more inline comment, but I think we should be good to go after that is addressed.

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java
Comment (PacoVK) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java]: Good point, have overseen this and corrected it

**Review by mposolda — APPROVED** [review]
@PacoVK Thanks a lot for your work on this!

**Pull request merged**
